drake_ros_introspection/include/drake_ros_introspection/simulator_monitor_builder.h

```diff
@@ -0,0 +1,120 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include <drake/systems/framework/diagram.h>
+#include <drake/systems/framework/input_port.h>
+#include <drake/systems/framework/output_port.h>
+#include <drake/systems/framework/system.h>
+#include <drake/systems/framework/system_visitor.h>
+#include <drake_ros_introspection/probes/probe_interface.h>
+#include <drake_ros_introspection/rules/port_probing_rule.h>
+#include <drake_ros_introspection/simulator_monitor.h>
+
+namespace drake_ros_introspection {
+
+template <typename T>
+class SimulatorMonitorBuilder {
+ public:
+  rules::AbstractPortProbingRuleWriter<drake::systems::InputPort<T>> For(
+      std::function<bool(const drake::systems::InputPort<T>&)> matcher) {
+    blueprint_.input_port_probing_rules.push_back({std::move(matcher)});
+    return rules::AbstractPortProbingRuleWriter<drake::systems::InputPort<T>>(
+        blueprint_.input_port_probing_rules.back());
+  }
+
+  rules::AbstractPortProbingRuleWriter<drake::systems::OutputPort<T>> For(
+      std::function<bool(const drake::systems::OutputPort<T>&)> matcher) {
+    blueprint_.output_port_probing_rules.push_back({std::move(matcher)});
+    return rules::AbstractPortProbingRuleWriter<drake::systems::OutputPort<T>>(
+        blueprint_.output_port_probing_rules.back());
+  }
+
+  SimulatorMonitor<T> Build(const drake::systems::System<T>& system) {
+    Worker worker{blueprint_};
+    system.Accept(&worker);
+    return worker.YieldOutput();
+  }
+
+ private:
+  struct Blueprint {
+    using InputPortProbingRule =
+        rules::PortProbingRule<drake::systems::InputPort<T>>;
+    std::vector<InputPortProbingRule> input_port_probing_rules;
+
+    using OutputPortProbingRule =
+        rules::PortProbingRule<drake::systems::OutputPort<T>>;
+    std::vector<OutputPortProbingRule> output_port_probing_rules;
+  };
+
+  class Worker : public drake::systems::SystemVisitor<T> {
+   public:
+    explicit Worker(const Blueprint& blueprint) : blueprint_(blueprint) {}
+
+    void VisitSystem(const drake::systems::System<T>& system) override {
+      BuildInputPortProbes(system);
+      BuildOutputPortProbes(system);
+    }
+
+    void VisitDiagram(const drake::systems::Diagram<T>& diagram) override {
```

Comment:
VisitSystem + VisitDiagram and therefor BuildInputPortProbes + BuildOutputPortProbes never seem to be used. Are they essential?

Reply:
Those are called (indirectly) from `SimulatorMonitor::Build`. This inner `Worker` class implements the [`SystemVisitor`](https://github.com/RobotLocomotion/drake/blob/master/systems/framework/system_visitor.h#L14) pattern to recursively crawl the given diagram.